cmd/attest.go

```diff
@@ -0,0 +1,193 @@
+	"google.golang.org/protobuf/proto"
+)
+
+var (
+	key              string
+	nonce            []byte
+	teenonce         []byte
+	format           string
+	gceZone          string
+	gceProjectID     string
+	gceProjectNumber uint64
+	gceInstanceName  string
+	gceInstanceID    uint64
+)
+
+var attestationKeys = map[string]map[tpm2.Algorithm]func(rw io.ReadWriter) (*client.Key, error){
+	"AK": {
+		tpm2.AlgRSA: client.AttestationKeyRSA,
+		tpm2.AlgECC: client.AttestationKeyECC,
+	},
+	"gceAK": {
+		tpm2.AlgRSA: client.GceAttestationKeyRSA,
+		tpm2.AlgECC: client.GceAttestationKeyECC,
+	},
+}
+
+var attestCmd = &cobra.Command{
+	Use:   "attest",
+	Short: "Create a remote attestation report",
+	Long: `Gather information for remote attestation.
+
+The Attestation report contains a quote on all available PCR banks ,a way to validate 
+the quote, and a TCG Event Log (Linux only).
+
+Use --key to specify the type of attestation key. It can be gceAK for GCE attestation
+key or AK for a custom attestation key. By default it uses gceAK.
+--algo flag override the public key algorithm for attestation key. If not provided then
+by default rsa is used.
+--nonce attaches even length extra data to the report and guarantees a fresh quote.
+--teenonce attaches a 64 bytes extra data to the attestation report of hardware and 
+guarantees a fresh quote.
+--format flag specify the type of output file. It can be "binarypb" or "textproto".
+By default it uses binarypb.
+--output flag provides file name to store attestation report in the format specified
+by --format flag.
+`,
+	Args: cobra.NoArgs,
+	RunE: func(cmd *cobra.Command, args []string) error {
+
+		rwc, err := openTpm()
+		if err != nil {
+			return err
+		}
+		defer rwc.Close()
+
+		createFunc := attestationKeys[key][keyAlgo]
+
+		attestationKey, err := createFunc(rwc)
+		if err != nil {
+			return fmt.Errorf("failed to create attestation key: %v", err)
+		}
+		defer attestationKey.Close()
+
+		attestopts := client.AttestOpts{}
+		attestopts.Nonce = nonce
+		if len(teenonce) != 0 {
+			attestopts.TEENonce = teenonce
+			attestopts.TEEDevice, err = client.CreateSevSnpDevice()
+			if err != nil {
+				return fmt.Errorf("collecting TEE attestation report: %v", err)
+			}
+		}
+
+		if ExternalTPM != nil {
+			if elg, ok := ExternalTPM.(client.EventLogGetter); ok {
+				attestopts.TCGEventLog, err = elg.EventLog()
+				if err != nil {
+					return fmt.Errorf("failed to retrieve TCG Event Log: %w", err)
+				}
+			}
+		}
+
+		attestation, err := attestationKey.Attest(attestopts)
+		if err != nil {
+			return fmt.Errorf("failed to collect attestation report: %v", err)
+		}
+
+		if key == "gceAK" {
+			getMetadata()
+			addGceInstanceInfo(attestation)
+		}
+
+		if len(output) != 0 {
+			var out []byte
+			if format == "binarypb" {
+				out, err = proto.Marshal(attestation)
+				if err != nil {
+					return fmt.Errorf("failed to marshal attestation proto: %v", attestation)
+				}
+			} else {
```

Comment:
Explicitly check the string for "textproto". Then return an error for incorrect flags. This checker util func should go in `flags.go`.

Reply:
Marshaling is implemented in attest.go and Unmarshaling is implemented in debugVerify.go .
We have to check the value of the format flag just before doing marshal and unmarshal .
And implementation of `Marshal` and `Unmarshal` are not common. 


Also I have added  `if len(output) != 0 ...` this check to ensure that it will not print an attestation report on stdout if file name is not provided.
And planning to add the same check in dubugVerify.go while taking the file from `--input` .

Reply:
Please remove the `if len(output) != 0` check. We should allow users to pipe from stdout into a file or into another command.

Reply:
OK , I'll remove this check.